flink/src/main/java/org/apache/iceberg/flink/IcebergFilesCommitter.java

```diff
@@ -0,0 +1,166 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class IcebergFilesCommitter extends RichSinkFunction<DataFile> implements
+    CheckpointListener, CheckpointedFunction {
+
+  private static final long serialVersionUID = 1L;
+  private static final Logger LOG = LoggerFactory.getLogger(IcebergFilesCommitter.class);
+  private static final String MAX_COMMITTED_CHECKPOINT_ID = "flink.max-committed.checkpoint.id";
+
+  private static final FlinkCatalogFactory CATALOG_FACTORY = new FlinkCatalogFactory();
+
+  private final String path;
+  private final SerializableConfiguration conf;
+  private final ImmutableMap<String, String> options;
+
+  private transient long maxCommittedCheckpointId;
+  private transient NavigableMap<Long, List<DataFile>> dataFilesPerCheckpoint;
+  private transient List<DataFile> dataFilesOfCurrentCheckpoint;
+  private transient Table table;
+
+  // State for all checkpoints;
+  private static final ListStateDescriptor<byte[]> STATE_DESCRIPTOR =
+      new ListStateDescriptor<>("checkpoints-state", BytePrimitiveArraySerializer.INSTANCE);
+  private transient ListState<byte[]> checkpointsState;
+
+  IcebergFilesCommitter(String path, Map<String, String> options, Configuration conf) {
+    this.path = path;
+    this.options = ImmutableMap.copyOf(options);
+    this.conf = new SerializableConfiguration(conf);
+  }
+
+  @Override
+  public void initializeState(FunctionInitializationContext context) throws Exception {
+    Catalog icebergCatalog = CATALOG_FACTORY.buildIcebergCatalog(path, options, conf.get());
+    table = icebergCatalog.loadTable(TableIdentifier.parse(path));
+    maxCommittedCheckpointId = parseMaxCommittedCheckpointId(table.currentSnapshot());
+
+    dataFilesPerCheckpoint = Maps.newTreeMap();
+    dataFilesOfCurrentCheckpoint = Lists.newArrayList();
+
+    checkpointsState = context.getOperatorStateStore().getListState(STATE_DESCRIPTOR);
+    if (context.isRestored()) {
+      dataFilesPerCheckpoint = deserializeState(checkpointsState.get().iterator().next());
+    }
+  }
+
+  @Override
+  public void snapshotState(FunctionSnapshotContext context) throws Exception {
+    long checkpointId = context.getCheckpointId();
+    LOG.info("Start to flush snapshot state to state backend, table: {}, checkpointId: {}", table, checkpointId);
+
+    // Update the checkpoint state.
+    dataFilesPerCheckpoint.put(checkpointId, ImmutableList.copyOf(dataFilesOfCurrentCheckpoint));
+
+    // Reset the snapshot state to the latest state.
+    checkpointsState.clear();
+    checkpointsState.addAll(ImmutableList.of(serializeState(dataFilesPerCheckpoint)));
+
+    // Clear the local buffer for current checkpoint.
+    dataFilesOfCurrentCheckpoint.clear();
+  }
+
+  @Override
+  public void notifyCheckpointComplete(long checkpointId) {
+    NavigableMap<Long, List<DataFile>> pendingFileMap = dataFilesPerCheckpoint.tailMap(maxCommittedCheckpointId, false);
```

Comment:
Why exclude `maxCommittedCheckpointId` and `checkpointId`? Just `dataFilesPerCheckpoint.tailMap(checkpointId, true)`? I don't understand why we need store previous checkpoint id.

Reply:
I've added few comments about those members here https://github.com/apache/iceberg/pull/1185/commits/6b43ec5b4183d708920e07a4378b8cb6a8ad290b.   

> Why exclude maxCommittedCheckpointId and checkpointId?

The `maxCommittedCheckPointId` 's data files have been committed to iceberg,  so no need to commit them again, otherwise the table will have redundant datas.  For current `checkpointId`,  we've flushed the `dataFilesOfCurrentCheckpoint` into `dataFilesPerCheckpoint` in `snapshotState` method,  so we did not exclude the files for current checkpointId.

Reply:
According to the current code, the data of the current checkpoint needs to be committed in the next cycle.
This should not be the ideal way, which can lead to excessive delay.

```
if (maxCommittedCheckpointId >= checkpointId) {
  return;
}
Map<Long, List<DataFile>> pendingFileMap = dataFilesPerCheckpoint.tailMap(checkpointId, true);
List<DataFile> pendingDataFiles = Lists.newArrayList();
for (List<DataFile> dataFiles : pendingFileMap.values()) {
      pendingDataFiles.addAll(dataFiles);
}
-- Do AppendFiles committing
maxCommittedCheckpointId = checkpointId;
pendingFileMap.clear();
```
This is the process I imagined.

Reply:
> the data of the current checkpoint needs to be committed in the next cycle.

That's not correct. Since we've merged the `dataFilesOfCurrentCheckpoint` into `dataFilesPerCheckpoint`,  so the  `dataFilesPerCheckpoint.tailMap(maxCommittedCheckpointId, false);`  will contains all the uncommitted data files,  include the files from current checkpoint.

About your described process: 

```java
Map<Long, List<DataFile>> pendingFileMap = dataFilesPerCheckpoint.tailMap(checkpointId, true);
```

This seems won't  commit all the previous files which have committed with failure .  How about the data files between (maxCommittedCheckpointId, checkpointId) ? 

In my thought,  the current process should be correct.  If there're some places I misunderstood, pls correct me,thanks.

Reply:
Sorry, wrong reading. Should be`headMap` instead of `tailMap`. You can think again about my above example.

I got your point about `tailMap`. But it is incorrect too. Consider the case:
- 1.snapshot(cpId)
- 2.snapshot(cpId + 1)
- 3.cpId success, notifyComplete(cpId)
- 4.cpId + 1 success, notifyComplete(cpId + 1)

If there is a failure between 3 and 4, the data in table will be incorrect.

Reply:
That's a great case (which I did not consider) to proof your point. Yes, we need the `headMap(checkpointId, true)` to avoid future data files being committed to iceberg table,  I've fixed this point in https://github.com/apache/iceberg/pull/1185/commits/598f2353ecad5eda730a1dda96275e552657e466.